File: 3734.lexicographically-smallest-palindromic-permutation-greater-than-target.cpp
#
# @lc app=leetcode id=3734 lang=cpp
#
# [3734] Lexicographically Smallest Palindromic Permutation Greater Than Target
#
# @lc code=start
class Solution {
public:
    string lexPalindromicPermutation(string s, string target) {
        int n = s.length();
        std::array<int, 26> freq = {0};
        for (char c : s) {
            ++freq[c - 'a'];
        }
        std::vector<char> odds;
        for (int i = 0; i < 26; ++i) {
            if (freq[i] % 2 == 1) {
                odds.push_back('a' + i);
            }
        }
        bool odd_length = (n % 2 == 1);
        if ((odd_length && odds.size() != 1) || (!odd_length && odds.size() != 0)) {
            return "";
        }
        char middle = odd_length ? odds[0] : '\0';
        int half = n / 2;
        std::array<int, 26> pairs = {0};
        for (int i = 0; i < 26; ++i) {
            pairs[i] = freq[i] / 2;
        }
        auto can_exceed = [&](int pos, const std::vector<char>& lft, const std::array<int, 26>& rem) -> bool {
            int midpos = n / 2;
            bool hasmid = odd_length;
            char midc = middle;
            for (int d = 0; d < n; ++d) {
                char pd = '\0';
                bool det = false;
                if (d < static_cast<int>(lft.size())) {
                    pd = lft[d];
                    det = true;
                } else if (d >= n - 1 - pos) {
                    int j = n - 1 - d;
                    pd = lft[j];
                    det = true;
                } else if (hasmid && d == midpos) {
                    pd = midc;
                    det = true;
                }
                if (!det) {
                    int rlen = half - (pos + 1);
                    std::vector<char> rlist;
                    for (int i = 0; i < 26; ++i) {
                        rlist.insert(rlist.end(), rem[i], static_cast<char>('a' + i));
                    }
                    std::sort(rlist.rbegin(), rlist.rend());
                    std::vector<char> full = lft;
                    full.insert(full.end(), rlist.begin(), rlist.end());
                    std::string pmax(n, '\0');
                    for (int j = 0; j < half; ++j) {
                        pmax[j] = full[j];
                        pmax[n - 1 - j] = full[j];
                    }
                    if (hasmid) {
                        pmax[midpos] = midc;
                    }
                    return pmax > target;
                }
                if (pd != target[d]) {
                    return pd > target[d];
                }
            }
            return false;
        };
        std::vector<char> lhalf;
        std::array<int, 26> rem_pairs = pairs;
        bool success = true;
        for (int pos = 0; pos < half; ++pos) {
            bool found = false;
            for (char ch = 'a'; ch <= 'z'; ++ch) {
                int idx = ch - 'a';
                if (rem_pairs[idx] > 0) {
                    --rem_pairs[idx];
                    lhalf.push_back(ch);
                    if (can_exceed(pos, lhalf, rem_pairs)) {
                        found = true;
                        break;
                    }
                    ++rem_pairs[idx];
                    lhalf.pop_back();
                }
            }
            if (!found) {
                success = false;
                break;
            }
        }
        if (!success) {
            return "";
        }
        std::string result(n, '\0');
        for (int j = 0; j < half; ++j) {
            result[j] = lhalf[j];
            result[n - 1 - j] = lhalf[j];
        }
        if (odd_length) {
            result[n / 2] = middle;
        }
        if (result > target) {
            return result;
        } else {
            return "";
        }
    }
};
# @lc code=end

File: 3734.lexicographically-smallest-palindromic-permutation-greater-than-target.py
#
# @lc app=leetcode id=3734 lang=python3
#
# [3734] Lexicographically Smallest Palindromic Permutation Greater Than Target
#

# @lc code=start
from collections import Counter

class Solution:
    def lexPalindromicPermutation(self, s: str, target: str) -> str:
        n = len(s)
        freq = Counter(s)
        odds = sum(1 for cnt in freq.values() if cnt % 2 == 1)
        if (n % 2 == 0 and odds > 0) or (n % 2 == 1 and odds != 1):
            return ""
        half = (n + 1) // 2
        m = n // 2
        for k in range(half, -1, -1):
            temp_freq = freq.copy()
            left = []
            possible = True
            for i in range(k):
                c = target[i]
                is_mid = (n % 2 == 1 and i == m)
                need = 1 if is_mid else 2
                if temp_freq[c] < need:
                    possible = False
                    break
                temp_freq[c] -= need
                left.append(c)
            if not possible:
                continue
            cand_left = left[:]
            if k == half:
                if all(cnt == 0 for cnt in temp_freq.values()):
                    pal = [""] * n
                    for i in range(half):
                        pal[i] = cand_left[i]
                    for i in range(m):
                        pal[n - 1 - i] = pal[i]
                    pal_str = ''.join(pal)
                    if pal_str > target:
                        return pal_str
                continue
            is_mid_k = (n % 2 == 1 and k == m)
            need_k = 1 if is_mid_k else 2
            chosen = None
            for o in range(ord(target[k]) + 1, ord('z') + 1):
                c = chr(o)
                if temp_freq[c] >= need_k:
                    chosen = c
                    break
            if chosen is None:
                continue
            temp_freq[chosen] -= need_k
            cand_left.append(chosen)
            fill_ok = True
            for pos in range(k + 1, half):
                is_mid_pos = (n % 2 == 1 and pos == m)
                need_pos = 1 if is_mid_pos else 2
                found = None
                for o in range(ord('a'), ord('z') + 1):
                    cc = chr(o)
                    if temp_freq[cc] >= need_pos:
                        found = cc
                        break
                if found is None:
                    fill_ok = False
                    break
                temp_freq[found] -= need_pos
                cand_left.append(found)
            if not fill_ok:
                continue
            pal = [""] * n
            for i in range(half):
                pal[i] = cand_left[i]
            for i in range(m):
                pal[n - 1 - i] = pal[i]
            return ''.join(pal)
        return ""
# @lc code=end